INTMDB-804: [Terraform] Create new TF Data Lake Pipelines resource and data sources. (#1174)
mongodb/terraform-provider-mongodbatlas: website/docs/d/data_lake_pipeline.html.markdown

---
layout: "mongodbatlas"
page_title: "MongoDB Atlas: data_lake"
sidebar_current: "docs-mongodbatlas-resource-data-lake"
description: |-
    Describe a Data Lake Pipeline.
---

# Data Source: mongodbatlas_data_lake_pipeline

> `mongodbatlas_data_lake_pipeline` describe a Data Lake Pipeline.

Reviewer: ```suggestion
`mongodbatlas_data_lake_pipeline` describes a Data Lake Pipeline.
```

> -> **NOTE:** Groups and projects are synonymous terms. You may find group_id in the official documentation.

Reviewer: ```suggestion
-> **NOTE:** Groups and projects are synonymous terms. You may find `group_id` in the official documentation.
```

> ~> **IMPORTANT:** All arguments including the password will be stored in the raw state as plain-text. [Read more about sensitive data in state.](https://www.terraform.io/docs/state/sensitive-data.html)

Reviewer: ```suggestion
~> **IMPORTANT:** All arguments including the password will be stored in plain-text. [Read more about handling sensitive data in Terraform state.](https://www.terraform.io/docs/state/sensitive-data.html)
```

> ~> **IMPORTANT:** All arguments including the password will be stored in the raw state as plain-text. [Read more about sensitive data in state.](https://www.terraform.io/docs/state/sensitive-data.html)

Reviewer 2: Is this a general statement we are now including?  I ask because I don't think this datasource has any sensitive values.

Author: > Is this a general statement we are now including?

@Zuhairahmed Could you confirm if we should always include this message? I removed it for now since there is no sensitive information in this resource.

Reviewer 3: Isn't this also dependant on the user's Terraform setup - for example, if they have a remote state via Terraform Cloud, their data won't be stored locally.

More high-level question - what led us to add the disclaimer? If it's a general policy statement, I think we should move it to a central document rather than needing to include it in each example.

Also, happy to make a follow up ticket for our plans here - since this resource doesn't have sensitive information, better to do separately.

Reviewer 2: @colm-quinn good question.   It will always be in plain text but if they are using Terraform Enterprise or Cloud the state is stored centrally and hence more protected.   

But I am note sure when this got added.   I don't remember it so I think @Zuhairahmed this is from after you started?  We have always had messages like this is particularly risky areas, like DB users or API keys, but to Colm's point we could have it just centrally in the main provider doc page and include it on a page only when it's a specifically risky resource.  

Agree a follow up ticket is likely the best course here. Then we can make an intentional decision on this.  @colm-quinn shall we do that and add as a topic for the Prod+Eng meeting?

Reviewer 3: sgtm, let's follow up there!

    provider_name               = "GCP"
    provider_region_name        = "US_EAST_4"
    provider_instance_size_name = "M10"
    cloud_backup                = true   // enable cloud backup snapshots
    mongo_db_major_version      = "6.0"
  }

resource "mongodbatlas_data_lake_pipeline" "pipeline" {
      project_id       = mongodbatlas_project.projectTest.project_id
      name       = "DataLakePipelineName"
      sink {
        type = "DLS"
        partition_fields {
            name = "access"
            order = 0
        }
      }

      source {
        type = "ON_DEMAND_CPS"
        cluster_name = mongodbatlas_cluster.automated_backup_test.name
        database_name = "sample_airbnb"
        collection_name = "listingsAndReviews"
      }

      transformations {
              field = "test"
              type  = "EXCLUDE"
      }

      transformations {
              field = "test22"
              type  = "EXCLUDE"
      }
}

data "mongodbatlas_data_lake_pipeline" "pipelineDataSource" {
  project_id       = mongodbatlas_data_lake_pipeline.pipeline.project_id
  name             = mongodbatlas_data_lake_pipeline.pipeline.name
}
```

## Argument Reference

> * `project_id` - (Required) The unique ID for the project to create a data lake pipeline.

Reviewer: ```suggestion
* `project_id` - (Required) The unique ID for the project to create a Data Lake Pipeline.
```

* `name` - (Required) Name of the Atlas Data Lake Pipeline.

## Attributes Reference

In addition to all arguments above, the following attributes are exported:

* `id` -  Unique 24-hexadecimal digit string that identifies the Data Lake Pipeline.
* `created_date` - Timestamp that indicates when the Data Lake Pipeline was created.
* `last_updated_date` - Timestamp that indicates the last time that the Data Lake Pipeline was updated.
* `state` - State of this Data Lake Pipeline.
* `sink` - Ingestion destination of a Data Lake Pipeline.
  * `sink.0.type` - Type of ingestion destination of this Data Lake Pipeline.
  * `sink.0.provider` - Target cloud provider for this Data Lake Pipeline.
  * `sink.0.region` - Target cloud provider region for this Data Lake Pipeline. [Supported cloud provider regions](https://www.mongodb.com/docs/datalake/limitations).
  * `sink.0.partition_fields` - Ordered fields used to physically organize data in the destination.
    * `sink.0.partition_fields.#.name` - Human-readable label that identifies the field name used to partition data.

>     * `sink.0.partition_fields.#.order` - Sequence in which MongoDB Cloud slices the collection data to create partitions. The resource expresses this sequence starting with zero.

Reviewer: Maybe:

```suggestion
    * `sink.0.partition_fields.#.order` - Sequence in which MongoDB Atlas slices the collection data to create partitions. The resource expresses this sequence starting with zero.
```

* `source` - Ingestion Source of a Data Lake Pipeline.
  * `source.0.type` - Type of ingestion source of this Data Lake Pipeline.
  * `source.0.cluster_name` - Human-readable name that identifies the cluster.
  * `source.0.collection_name` - Human-readable name that identifies the collection.
  * `source.0.database_name` - Human-readable name that identifies the database.
  * `source.0.project_id` - Unique 24-hexadecimal character string that identifies the project.
  * `source.0.policyItemId` - Unique 24-hexadecimal character string that identifies a policy item.
* `transformations` - Fields to be excluded for this Data Lake Pipeline.
  * `transformations.#.field` - Key in the document.
  * `transformations.#.type` - Type of transformation applied during the export of the namespace in a Data Lake Pipeline.
* `snapshots` - List of backup snapshots that you can use to trigger an on demand pipeline run.
  * `snapshots.#.id` - Unique 24-hexadecimal digit string that identifies the snapshot.
  * `snapshots.#.provider` - Human-readable label that identifies the cloud provider that stores this snapshot.

>   * `snapshots.#.created_at` - Date and time when MongoDB Cloud took the snapshot.

Reviewer: ditto

Author: updated

---
layout: "mongodbatlas"
page_title: "MongoDB Atlas: data_lake"
sidebar_current: "docs-mongodbatlas-resource-data-lake"
description: |-
    Describe a Data Lake Pipeline.
---

# Data Source: mongodbatlas_data_lake_pipeline

`mongodbatlas_data_lake_pipeline` describe a Data Lake Pipeline.

-> **NOTE:** Groups and projects are synonymous terms. You may find group_id in the official documentation.

~> **IMPORTANT:** All arguments including the password will be stored in the raw state as plain-text. [Read more about sensitive data in state.](https://www.terraform.io/docs/state/sensitive-data.html)

## Example Usages

```terraform
resource "mongodbatlas_project" "projectTest" {
  name   = "NAME OF THE PROJECT"
  org_id = "ORGANIZATION ID"
}

resource "mongodbatlas_cluster" "automated_backup_test" {
    project_id   = "63f4d4a47baeac59406dc131"
    name         = "automated-backup-test"

    provider_name               = "GCP"
    provider_region_name        = "US_EAST_4"

>     provider_instance_size_name = "M30"

Reviewer 3: Same point on using an M10

    cloud_backup                = true   // enable cloud backup snapshots

>     mongo_db_major_version      = "4.4"

Reviewer 4: ```suggestion
    mongo_db_major_version      = "6.0"
```

>     disk_size_gb = "350"

Reviewer 4: ```suggestion
```
